generator/src/main/java/com/scottlogic/deg/generator/validators/ContradictionTreeValidator.java

```diff
@@ -0,0 +1,145 @@
+package com.scottlogic.deg.generator.validators;
+
+import com.google.inject.Inject;
+import com.scottlogic.deg.common.profile.Field;
+import com.scottlogic.deg.common.profile.ProfileFields;
+import com.scottlogic.deg.generator.decisiontree.ConstraintNode;
+import com.scottlogic.deg.generator.decisiontree.DecisionNode;
+import com.scottlogic.deg.generator.decisiontree.DecisionTree;
+import com.scottlogic.deg.generator.decisiontree.Node;
+import com.scottlogic.deg.generator.fieldspecs.FieldSpec;
+import com.scottlogic.deg.generator.generation.DataGeneratorMonitor;
+import com.scottlogic.deg.generator.walker.reductive.ReductiveTreePruner;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+
+public class ContradictionTreeValidator {
+    private final ContradictionChecker contradictionChecker;
+    private final ReductiveTreePruner treePruner;
+
+    @Inject
+    public ContradictionTreeValidator(ContradictionChecker contradictionChecker, ReductiveTreePruner treePruner){
+        this.contradictionChecker = contradictionChecker;
+        this.treePruner = treePruner;
+    }
+
+    public DecisionTree reportThenCullContradictions(DecisionTree decisionTree, DataGeneratorMonitor monitor) {
+        Collection<Node> contradictingNodes = reportContradictions(decisionTree);
+        if (contradictingNodes.contains(decisionTree.getRootNode())) {
+            monitor.addLineToPrintAtEndOfGeneration(
+                "The provided profile is wholly contradictory. No fields can successfully be fixed.",
+                System.err
+            );
+            return new DecisionTree(null, decisionTree.getFields(), decisionTree.getDescription());
+        } else if (contradictingNodes.size() > 0) {
+            monitor.addLineToPrintAtEndOfGeneration(
+                "Warning: There are " +
+                    contradictingNodes.size() +
+                    " partial contradiction(s) in the profile. Run the profile through the visualiser for more information.",
+                System.err
+            );
+
+            // Remove all contradicting subtrees in the decision tree and return the new tree.
+            ConstraintNode prunedRootNode = pruneOutPartialContradictions(decisionTree.getRootNode(), decisionTree.getFields());
+            return new DecisionTree(prunedRootNode, decisionTree.getFields(), decisionTree.getDescription());
+        }
+        // No contradictions.
+        return decisionTree;
```

Comment:
I'm not sure it would read as well. It currently reads:
```
if (contradictingNodes.contains(decisionTree.getRootNode())) {
  //...
} else if (contradictingNodes.size() > 0) {
  //...
}
return decisionTree;
```
But with change would read:
```
if (!contradictingNodes.contains(decisionTree.getRootNode()) && contradictingNodes.isEmpty()) {
  return decisionTree;
} else if (contradictingNodes.contains(decisionTree.getRootNode())) {
  //...
}
//...
```

Reply:
it would just be

`contradictingnodes.isEmpty()`

if it contains the root node, it's not going to be empty :P